src/passes/OptimizeCasts.cpp

```diff
@@ -102,6 +120,273 @@ namespace wasm {
+  // For some index, transfers the RefCasts that can currently be
+  // moved to the refCastToApply result index and resets the earliest
+  // reachable local.get to a nullptr. Done when we visit an expression
+  // where we cannot move casts past.
+  void updateRefCastMoves(size_t index) {
+    if (earliestRefCastReachable[index]) {
+      auto& applyVector = refCastToApply[earliestRefCastReachable[index]];
+      for (size_t i = 0; i < movedRefCast[index].size(); i++) {
+        if (!movedRefCast[index][i].atTarget) {
+          applyVector.push_back(movedRefCast[index][i]);
+        }
+      }
+      movedRefCast[index].clear();
+      earliestRefCastReachable[index] = nullptr;
+    }
+  }
+
+  // Same as above but for RefAs.
+  void updateRefAsMoves(size_t index) {
+    if (earliestRefAsReachable[index]) {
+      auto& applyVector = refAsToApply[earliestRefAsReachable[index]];
+      for (size_t i = 0; i < movedRefAs[index].size(); i++) {
+        if (!movedRefAs[index][i].atTarget) {
+          applyVector.push_back(movedRefAs[index][i]);
+        }
+      }
+      movedRefAs[index].clear();
+      earliestRefAsReachable[index] = nullptr;
+    }
+  }
+
+  static void doNoteNonLinear(EarlyCastFinder* self, Expression**) {
+    for (size_t i = 0; i < self->numLocals; i++) {
+      self->updateRefCastMoves(i);
+      self->updateRefAsMoves(i);
+    }
+  }
+
+  void visitExpression(Expression* curr) {
+    // A new one is instantiated for each expression to determine
+    // if a cast can be moved past it.
+    ShallowEffectAnalyzer currAnalyzer(options, *getModule(), curr);
+
+    if (testRefCast.invalidates(currAnalyzer)) {
+      for (size_t i = 0; i < numLocals; i++) {
+        updateRefCastMoves(i);
+      }
+    }
+
+    if (testRefAs.invalidates(currAnalyzer)) {
+      for (size_t i = 0; i < numLocals; i++) {
+        updateRefAsMoves(i);
+      }
+    }
+  }
+
+  void visitLocalSet(LocalSet* curr) {
+    visitExpression(curr);
+    updateRefCastMoves(curr->index);
+    updateRefAsMoves(curr->index);
+  }
+
+  void visitLocalGet(LocalGet* curr) {
+    visitExpression(curr);
+
+    if (!earliestRefCastReachable[curr->index]) {
+      earliestRefCastReachable[curr->index] = curr;
+    }
+
+    if (!earliestRefAsReachable[curr->index]) {
+      earliestRefAsReachable[curr->index] = curr;
+    }
+  }
+
+  void visitRefAs(RefAs* curr) {
+    visitExpression(curr);
+
+    auto* fallthrough = Properties::getFallthrough(curr, options, *getModule());
+
+    if (auto* get = fallthrough->dynCast<LocalGet>()) {
+      if (get != earliestRefAsReachable[get->index]) {
+        bool alreadyAdded = false;
+        for (size_t i = 0; i < movedRefAs[get->index].size(); i++) {
+          if (movedRefAs[get->index][i].op == curr->op) {
+            alreadyAdded = true;
+            break;
+          }
+        }
+
+        if (!alreadyAdded) {
+          movedRefAs[get->index].emplace_back(
+            curr->op, get == earliestRefAsReachable[get->index]);
```

Comment:
Why doesn't this place add to the final output place, `refAsToApply`? I'm having a hard time understanding why it is queued here and processed later. Sorry if I'm missing something obvious.

(I guess queuing can help by letting us decide later what is more optimal. But also updating `refAsToApply` to something better if we find it seems equally good.)

Reply:
You're right, it could be added immediately. I just followed the pattern in `visitRefCast`, where I queue it because there may be more refined casts later.  But for  a `RefAs`, since the operators are orthogonal to each other and neither is "better", this doesn't matter.

Reply:
Can't we do the same for visitRefCast? Again, if we see a more refined cast later can't be update the final output place as we go?

I mean, if we see
1. A LocalGet
2. A cast to $B
3. A cast to $C

Then at step 3 we can overwrite what step 2 wrote.